Objects/moduleobject.c

```diff
@@ -784,11 +786,69 @@ _PyModuleSpec_IsUninitializedSubmodule(PyObject *spec, PyObject *name)
     return rc;
 }
 
+static int
+_is_module_possibly_shadowing(PyObject *origin)
+{
+    // origin must be a unicode subtype
+    // Returns 1 if the module at origin could be shadowing a module of the
+    // same name later in the module search path. The condition we check is basically:
+    // not sys.flags.safe_path and os.path.dirname(origin) == (sys.path[0] or os.getcwd())
+    // Returns 0 otherwise (or if we aren't sure)
+    // Returns -1 if an error occurred that should be propagated
+    if (origin == NULL) {
+        return 0;
+    }
+    const PyConfig *config = _Py_GetConfig();
+    if (config->safe_path) {
+        return 0;
+    }
+
+    PyObject *sys_path = PySys_GetObject("path");
+    if (sys_path == NULL || !PyList_Check(sys_path) || PyList_GET_SIZE(sys_path) == 0) {
+        return 0;
+    }
+    PyObject *py_sys_path_0 = PyList_GET_ITEM(sys_path, 0);
+    if (!PyUnicode_Check(py_sys_path_0)) {
+        return 0;
+    }
+    wchar_t sys_path_0[MAXPATHLEN + 1];
+    Py_ssize_t size = PyUnicode_AsWideChar(py_sys_path_0, sys_path_0, MAXPATHLEN);
+    if (size < 0) {
+        return -1;
+    }
+    assert(size <= MAXPATHLEN);
+    sys_path_0[size] = L'\0';
+    if (size == 0) {
+        // if sys.path[0] == "", treat it as if it were the current directory
+        if (!_Py_wgetcwd(sys_path_0, MAXPATHLEN)) {
+            return -1;
+        }
+    }
```

Comment:
Why not use `config->module_search_paths` or even `config->sys_path_0`?  Both should have the info we want, already a `wchar_t *`.  As a bonus, this avoids having to deal with any user modifications to `sys.path`.

```C
    wchar_t *sys_path_0 = config->sys_path_0;
    if (sys_path_0 == NULL) {
        return 0;
    }
```

or maybe

```C
    if (!config->module_search_paths_set) {
        return 0;
    }
    assert(config->module_search_paths->length > 0);
    wchar_t *sys_path_0 = config->module_search_paths.items[0];
    assert(sys_path_0 != NULL);
```

It would probably make sense to combine those, using `config->module_search_paths` as a fallback:

```suggestion
    wchar_t *sys_path_0 = config->sys_path_0;
    if (sys_path_0 == NULL) {
        if (!config->module_search_paths_set || config->module_search_paths.length == 0) {
            return 0;
        }
        sys_path_0 = config->module_search_paths.items[0];
        assert(sys_path_0 != NULL);
    }
```

If both are set, `config->sys_path_0` should always match `config->module_search_paths.items[0]`.  However, that constraint should be enforced elsewhere already (e.g. Python/pathconfig.c), so there's no need to check here.

Reply:
We could also fall back to duplicating the relevant parts of `pymain_run_python()` in Modules/main.c, including calling `_PyPathConfig_ComputeSysPath0()`.  However, that shouldn't be necessary.  It would only benefit embedders and they would likely want more than part of `pymain_run_python()` duplicated.

Reply:
Thanks for the review! I've incorporated the suggestion — it's nice that we handle user modifications to sys.path, and I've added a test for this. Note we still need to do some cwd munging in the case `config->sys_path_0` is the empty str.